python_scripts/load_arrow.py

```diff
@@ -0,0 +1,348 @@
+        if too_long_aliases:
+            raise ValueError(
+                "Aliases file contains aliases longer than the allowed length (32)"
+            )
+        return aliases
+
+    def convert_int64(self):
+        """
+        Stata can only deal with int8, int16, int32; if we have ints that require
+        int64, we need to convert them to float
+        First we attempt to convert to int32; if that doesn't raise an exception,
+        we can leave the column as it is
+        """
+        for column_name in self.column_names:
+            if is_int64(self.arrow_table.schema.field(column_name).type):
+                try:
+                    converted = self.arrow_table[column_name].cast(int32())
+                except Exception:
+                    converted = self.arrow_table[column_name].cast(float64())
+                    self._replace_column(column_name, converted)
+
+    def get_column_types(self):
+        """
+        Method to return the modal dtype across columns of the data frames
+        :return: A dictionary containing the modal type for each column
+        """
+        column_test_mapping = {
+            is_floating: "float",
+            is_boolean: "boolean",
+            is_timestamp: "timestamp",
+            is_date: "date",
+            is_integer: "int",
+            is_dictionary: "category",
+        }
+        column_types = {}
+        for column_name in self.column_names:
+            column_metadata = self.arrow_table.schema.field(column_name)
+            column_types[column_name] = next(
+                (
+                    col_type
+                    for test_fn, col_type in column_test_mapping.items()
+                    if test_fn(column_metadata.type)
+                ),
+                "string",
+            )
+        return column_types
+
+    def clean_names(self):
+        """
+        Function to ensure all of the names of the variables in the table
+        are valid Stata variable names.
+        If an aliases file has been provided, use it to rename the columns.
+        """
+        # Creates a mapping from existing column names to names allowed by Stata
+        clean_column_names = []
+        # Keep a note of names that are too long so we can report those back to the user.
+        too_long_names = []
+        for varname in self.arrow_table.column_names:
+            if len(varname) > 32 and varname not in self.aliases:
+                too_long_names.append(varname)
+                continue
+            # Get the aliases name, if one exists, and run it through the sfitoolkit method to make
+            # sure it's valid for stata
+            cleaned_name = sfi.SFIToolkit.strToName(
+                self.aliases.get(varname, varname), prefix=False
+            )
+            if cleaned_name != varname:
+                print(f"{varname} aliased to {cleaned_name}")
+            clean_column_names.append(cleaned_name)
+
+        if too_long_names:
```

Comment:
If the too_long_names can be automatically shortened w/o clashing, should we do that? And print a warning that we have done it?

I assume stata users are used to working withing the 32 character limit anyway? Is there some form of autoshortening that stata provides when loading from csv we can copy? e.g. "`too_long_name_xxxxx" -> "to_long_na~1"` or similar?

Reply:
They could be automatically shortened, but we decided not to, and raising exceptions with warnings was the better route - see [this slack thread](https://bennettoxford.slack.com/archives/C03FB777L1M/p1681229679397389) that I totally forgot was needed for context